### ISSUE TITLE
Ascent from binary cache not fixing path correctly, can cause runtime failure

### ISSUE THREAD

**Issue by EthanS94** [LLNL/spack#21739]
### Steps to reproduce the issue

With access to summit, this can be reproduced using workflow in https://github.com/cinemascienceworkflows/2020-10_Nyx . I can help with running this if needed.

```console
$ git clone https://github.com/cinemascienceworkflows/2020-10_Nyx.git
$ cd 2020-10_Nyx
$ vi bootstrap.env # Set WarningNotSet to appropriate variables
$ ./execute
```

### Error Message

The workflow pulls ascent and its dependencies from E4S cache, then builds Nyx using that ascent. The Nyx build is successful, but fails at runtime with: 
```console
$ vi <install_location>/pantheon/ECP-Examples_Nyx-002/results/Nyx64_Test.744775
...
[Error] Ascent::execute
file: /tmp/lpeyrala/spack-stage/spack-stage-ascent-0.6.0-bpoqtmc5f6kjhxeto2e6xqynicfrorpr/spack-src/src/ascent/utils/ascent_web_interface.cpp
line: 100
message:
Web client root directory (/dev/shm/pantheon-spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholde/linux-rhel7-power9le/gcc-6.4.0/ascent-0.6.0-bpoqtmc5f6kjhxeto2e6xqynicfrorpr/share/ascent/web_clients)  is missing

User defined signal 2
```

I could be wrong, but I think the problem is this path:
`/dev/shm/pantheon-spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholde/linux-rhel7-power9le/gcc-6.4.0/ascent-0.6.0-bpoqtmc5f6kjhxeto2e6xqynicfrorpr/share/ascent/web_clients`

Isn't being rewritten to the correct path, which for me would be:
`/gpfs/alpine/csc420/scratch/stam/pantheon/ECP-Examples_Nyx-002/spack/opt/spack/linux-rhel7-power9le/gcc-6.4.0/ascent-0.6.0-bpoqtmc5f6kjhxeto2e6xqynicfrorpr/share/ascent/web_clients`

### Information on your system

```console
$ spack debug report
* **Spack:** 0.16.0-6-676d68a
* **Python:** 2.7.5
* **Platform:** linux-rhel7-power9le
```

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

**Comment by EthanS94** [comment]
@cyrush I think you are aware of this issue. Let me know if you are not the right person to ping on this. Thanks!

**Comment by cyrush** [comment]
@EthanS94 yes, my understanding is that this is a limitation with spack's binary relocation mechanism. 

We can create an option in ascent that will give you the power to set the path yourself, but ideally we want spack to handle this case since it is re-wiring the paths in the binaries.

**Comment by EthanS94** [comment]
@cyrush 

> We can create an option in ascent that will give you the power to set the path yourself, but ideally we want spack to handle this case since it is re-wiring the paths in the binaries.

Can you share how to do this? We could implement this as a spack patch until this issue is fixed.

**Comment by EthanS94** [comment]
Adding @dhrogers to follow this

**Comment by cyrush** [comment]
To accomplish this: we would have to add a new feature to a future version of ascent that allows folks to hand is a path via the api, instead of relying on the built-in path which isn't properly relocated.

**Comment by dhrogers** [comment]
Is there a robust workaround? We'd like to demo these workflows, with E4S cache and Spack builds at the April ECP meeting.

**Comment by cyrush** [comment]
I can get the option in for you well before then, hopefully spack folks can dive into the relocation issue in parallel.

**Comment by dhrogers** [comment]
Thanks.

**Comment by tgamblin** [comment]
This should be fixable, as it's relocating from a long path to a short path.  I am not sure why we do not catch this -- we do relocate strings in binaries in this case.

**Comment by tgamblin** [comment]
@cyrush: is ascent linked as a static lib or a dynamic lib here?  We may not be handling the static lib case properly.

**Comment by cyrush** [comment]
@tgamblin -- combed though the setup and verified that this case is static libs (b/c of cuda)

**Comment by tgamblin** [comment]
Lemme get someone to dig into this one -- we should be able to handle this case, but I checked and we don't currently recognize that static libs need relocation.  It's a bit more complicated than that b/c they need one type of relocation we do for dynamic libs (strings) but not the other (RPATHs) so it'll take a bit of tweaking, but I think we already have the code, just not the glue.

**Comment by alalazo** [comment]
I think this would be solved by a finer analysis of the binaries to check whether we need RPATH relocation, string shortening or both. So essentially an improvement over the current state of build caches.
